apps/server/src/modules/tool/uc/external-tool-configuration.uc.ts

```diff
@@ -31,7 +42,47 @@ export class ExternalToolConfigurationUc {
 		return availableTools;
 	}
 
-	async getExternalToolForSchool(
+	public async getAvailableToolsForContext(
```

Comment:
You can divide this function and can use parallel calls to get the tools for a better performance. Something like this:

```
private async getSchoolAndContextTools(schoolId: EntityId, contextId: EntityId): Promise<[SchoolExternalToolDO[], ContextExternalToolDO[]]> {
        const [schoolExternalTools, contextToolsInUse] = await Promise.all([
            this.schoolExternalToolService.findSchoolExternalTools({ schoolId }),
            this.contextExternalToolService.findContextExternalTools({ contextId })
        ]);

        return [schoolExternalTools, contextToolsInUse];
    }

    private getSchoolToolsInUse(schoolExternalTools: SchoolExternalToolDO[], contextToolsInUse: ContextExternalToolDO[]): SchoolExternalToolDO[] {
        return schoolExternalTools.filter(
            (schoolExternalTool) => contextToolsInUse.some((contextExternalTool) => contextExternalTool.schoolToolId === schoolExternalTool.id)
        );
    }

    private getAvailableTools(externalTools: ExternalToolDO[], schoolExternalTools: SchoolExternalToolDO[], toolIdsInUse: EntityId[]): ExternalToolDO[] {
        return externalTools.filter((tool) => {
            const hasSchoolTool = schoolExternalTools.some((schoolExternalTool) => schoolExternalTool.toolId === tool.id);

            return !tool.isHidden && !!tool.id && !toolIdsInUse.includes(tool.id) && hasSchoolTool;
        });
    }

    public async getAvailableToolsForContext(
        userId: EntityId,
        schoolId: EntityId,
        contextId: EntityId,
        contextType: ToolContextType
    ): Promise<ExternalToolDO[]> {
        await this.ensureContextPermission(userId, contextId, contextType);

        const [schoolExternalTools, contextToolsInUse] = await this.getSchoolAndContextTools(schoolId, contextId);

        const schoolToolsInUse = this.getSchoolToolsInUse(schoolExternalTools, contextToolsInUse);

        const toolIdsInUse = schoolToolsInUse.map((schoolExternalTool) => schoolExternalTool.toolId);

        const externalTools = await this.externalToolService.findExternalTools({}).then(res => res.data);

        return this.getAvailableTools(externalTools, schoolExternalTools, toolIdsInUse);
    }
```

Reply:
Did you say

> better performance

and wrote JavaScript code? 😆 

Btw did ChatGPT write this?
```
const externalTools = await this.externalToolService.findExternalTools({}).then(res => res.data);

return this.getAvailableTools(externalTools, schoolExternalTools, toolIdsInUse);
```
does not look like code we would write...